File: src/main/java/net/lumalyte/trivia/listeners/ChatListener.java
package net.lumalyte.trivia.listeners;

import net.lumalyte.trivia.TriviaPlugin;
import net.lumalyte.trivia.managers.TriviaManager;
import net.lumalyte.trivia.util.MessageUtil;
import io.papermc.paper.event.player.AsyncChatEvent;
import net.kyori.adventure.text.TextComponent;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.AsyncPlayerChatEvent;

public class ChatListener implements Listener {
    private final TriviaPlugin plugin;
    private final TriviaManager triviaManager;

    public ChatListener(TriviaPlugin plugin, TriviaManager triviaManager) {
        this.plugin = plugin;
        this.triviaManager = triviaManager;
        
        // Register events manually to ensure we get priority
        plugin.getServer().getPluginManager().registerEvent(
            AsyncPlayerChatEvent.class,
            this,
            EventPriority.LOWEST,
            (listener, event) -> onLegacyChat((AsyncPlayerChatEvent) event),
            plugin,
            true
        );
        
        plugin.getServer().getPluginManager().registerEvent(
            AsyncChatEvent.class,
            this,
            EventPriority.LOWEST,
            (listener, event) -> onModernChat((AsyncChatEvent) event),
            plugin,
            true
        );
    }

    private void onLegacyChat(AsyncPlayerChatEvent event) {
        if (triviaManager.handleChat(event)) {
            event.setCancelled(true);
            String muteMsg = plugin.getConfig().getString("game.mute-incorrect.message",
                "&cYou are muted until this trivia game ends!");
            MessageUtil.sendMessage(event.getPlayer(), muteMsg);
            return;
        }

        // Get and normalize the message
        String originalMessage = MessageUtil.stripColor(event.getMessage().trim().toLowerCase());
        String answer = originalMessage;

        // Check if it's a valid answer format
        boolean isValidAnswer = false;
        
        // Check for multiple choice (A, B, C, D)
        if (answer.length() == 1 && answer.matches("[abcd]")) {
            isValidAnswer = true;
        }

        // Check for true/false (T, F, True, False)
        if (answer.matches("^(t(rue)?|f(alse)?)$")) {
            // Normalize to single letter
            answer = answer.startsWith("t") ? "a" : "b";
            isValidAnswer = true;
        }
        
        if (isValidAnswer) {
            // Store final answer for lambda
            final String finalAnswer = answer;
            
            // Cancel the chat message
            event.setCancelled(true);

            // Process the answer on the main thread
            plugin.getServer().getScheduler().runTask(plugin, () -> 
                triviaManager.checkAnswer(event.getPlayer(), finalAnswer));
        }
    }

    private void onModernChat(AsyncChatEvent event) {
        if (triviaManager.handleChat(event)) {
            event.setCancelled(true);
            String muteMsg = plugin.getConfig().getString("game.mute-incorrect.message",
                "&cYou are muted until this trivia game ends!");
            MessageUtil.sendMessage(event.getPlayer(), muteMsg);
            return;
        }

        // Get message content
        if (!(event.message() instanceof TextComponent textComponent)) {
            return;
        }

        // Get and normalize the message
        String originalMessage = MessageUtil.stripColor(textComponent.content().trim().toLowerCase());
        String answer = originalMessage;

        // Check if it's a valid answer format
        boolean isValidAnswer = false;
        
        // Check for multiple choice (A, B, C, D)
        if (answer.length() == 1 && answer.matches("[abcd]")) {
            isValidAnswer = true;
        }

        // Check for true/false (T, F, True, False)
        if (answer.matches("^(t(rue)?|f(alse)?)$")) {
            // Normalize to single letter
            answer = answer.startsWith("t") ? "a" : "b";
            isValidAnswer = true;
        }
        
        if (isValidAnswer) {
            // Store final answer for lambda
            final String finalAnswer = answer;
            
            // Cancel the chat message
            event.setCancelled(true);

            // Process the answer on the main thread
            plugin.getServer().getScheduler().runTask(plugin, () -> 
                triviaManager.checkAnswer(event.getPlayer(), finalAnswer));
        }
    }
} 

File: src/main/kotlin/net/badgersmc/trivia/infrastructure/bukkit/ChatListener.kt
package net.badgersmc.trivia.infrastructure.bukkit

import net.badgersmc.nexus.i18n.LangService
import net.badgersmc.trivia.application.ChatPlatform
import net.badgersmc.trivia.application.TriviaService
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.player.AsyncPlayerChatEvent

/**
 * Single listener for mute enforcement and answer parsing (REQ-019).
 * Uses Bukkit's AsyncPlayerChatEvent (not Paper's AsyncChatEvent) to avoid
 * Component-type mismatches and priority conflicts with RoseChat.
 *
 * On vanilla Paper: cancels AsyncPlayerChatEvent at LOWEST priority.
 * On RoseChat: mute only enforced for the trivia channel; other channels are untouched.
 */
class ChatListener(
    private val triviaService: TriviaService,
    private val lang: LangService,
    private val chatPlatform: ChatPlatform,
) : Listener {

    @EventHandler(priority = EventPriority.LOWEST, ignoreCancelled = false)
    fun onChat(event: AsyncPlayerChatEvent) {
        val player = event.player

        // Only active during a game
        if (!triviaService.isActive) return

        val content = event.message.trim()
        if (content.isEmpty()) return

        // Channel/platform check first — only trivia-channel messages proceed
        if (!chatPlatform.isAnswerChat(player, content)) return
        val answer = chatPlatform.extractAnswer(content) ?: return

        // Mute enforcement — only blocks trivia-channel answers
        if (chatPlatform.isMuted(player) && !player.hasPermission("lumatrivia.mute.bypass")) {
            event.isCancelled = true
            player.sendMessage(lang.msg("mute.muted"))
            return
        }

        // Valid answer formats: single letter or t/f/true/false
        val normalized = answer.trim().lowercase()
        val question = triviaService.currentQuestion ?: return
        val maxLetter = 'a' + (question.answerCount - 1)
        val isValidAnswer = when {
            normalized.length == 1 && normalized[0] in 'a'..maxLetter -> true
            normalized.matches(Regex("^(t(rue)?|f(alse)?)$")) -> true
            else -> false
        }

        if (isValidAnswer) {
            event.isCancelled = true

            // Early-out: block duplicate answers before scheduling
            if (triviaService.hasPlayerAnswered(player.uniqueId)) {
                player.sendMessage(lang.msg("game.already_answered"))
                return
            }

            val mapped = when {
                normalized.startsWith("t") -> "true"
                normalized.startsWith("f") -> "false"
                else -> normalized
            }
            player.server.scheduler.runTask(
                player.server.pluginManager.getPlugin("LumaTrivia")!!,
                Runnable {
                    if (!player.isOnline) return@Runnable
                    val result = triviaService.checkAnswer(player, mapped)
                    when (result) {
                        TriviaService.AnswerResult.CORRECT -> {
                            val q = triviaService.currentQuestion ?: return@Runnable
                            player.server.broadcast(
                                lang.msg(
                                    "game.correct_answer",
                                    "player" to player.name,
                                    "answer" to q.correctAnswer,
                                    "letter" to q.correctAnswerLetter,
                                )
                            )
                        }
                        TriviaService.AnswerResult.WRONG -> {
                            player.server.broadcast(
                                lang.msg(
                                    "game.wrong_answer",
                                    "player" to player.name,
                                    "answer" to content,
                                )
                            )
                            if (chatPlatform.isMuted(player)) {
                                player.sendMessage(lang.msg("mute.muted"))
                            }
                        }
                        TriviaService.AnswerResult.ALREADY_ANSWERED -> {
                            player.sendMessage(lang.msg("game.already_answered"))
                        }
                        TriviaService.AnswerResult.NO_GAME -> {}
                    }
                }
            )
        } else {
            // Invalid format — cancel the message and give a hint
            event.isCancelled = true
            val hint = if (question.answerCount == 2 && question.type == "boolean") {
                lang.msg("game.hint.truefalse")
            } else {
                val last = 'a' + (question.answerCount - 1)
                lang.msg("game.hint.letters", "first" to "a", "last" to last.toString())
            }
            player.sendMessage(hint)
        }
    }
}
